# Stack Overflow Analysis Project 🧐

## Introduction

In this project we will be using our analytical skill to answer a few questions using the data from the Stack Overflow 2023 survey. 

**Questions**:
1. At what jobs get paid the most?
2. How much does remote working matter to employees?
3. How does coding experience affect the level of pay?
4. What's the most popular method of learning to code? 
5. Are the chances of getting a job higher as a developer if you have a master's degree?


## Data Set 📊 and Libraries 📚

In [81]:
# import the basic data sets
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [82]:
# call in the libraries
df = pd.read_csv('../data/survey_results_public.csv')
df2 = pd.read_csv('../data/survey_results_schema.csv')

### Looking at both data sets

In [83]:
pd.set_option('display.max_columns',84)
df.head()

ResponseId     Q120                      MainBranch              Age  \
0           1  I agree                   None of these  18-24 years old   
1           2  I agree  I am a developer by profession  25-34 years old   
2           3  I agree  I am a developer by profession  45-54 years old   
3           4  I agree  I am a developer by profession  25-34 years old   
4           5  I agree  I am a developer by profession  25-34 years old   

                                          Employment  \
0                                                NaN   
1                                Employed, full-time   
2                                Employed, full-time   
3                                Employed, full-time   
4  Employed, full-time;Independent contractor, fr...   

                             RemoteWork  \
0                                   NaN   
1                                Remote   
2  Hybrid (some remote, some in-person)   
3  Hybrid (some remote, some in-person)   
4                                Remote   

                                    CodingActivities  \
0                                                NaN   
1  Hobby;Contribute to open-source projects;Boots...   
2  Hobby;Professional development or self-paced l...   
3                                              Hobby   
4  Hobby;Contribute to open-source projects;Profe...   

                                        EdLevel  \
0                                           NaN   
1  Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
2  Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
3  Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
4  Bachelor’s degree (B.A., B.S., B.Eng., etc.)   

                                           LearnCode  \
0                                                NaN   
1  Books / Physical media;Colleague;Friend or fam...   
2  Books / Physical media;Colleague;On the job tr...   
3  Colleague;Friend or family member;Other online...   
4  Books / Physical media;Online Courses or Certi...   

                                     LearnCodeOnline  LearnCodeCoursesCert  \
0                                                NaN                   NaN   
1  Formal documentation provided by the owner of ...                 Other   
2  Formal documentation provided by the owner of ...                   NaN   
3  Formal documentation provided by the owner of ...                   NaN   
4  Formal documentation provided by the owner of ...  Other;Codecademy;edX   

  YearsCode YearsCodePro                               DevType  \
0       NaN          NaN                                   NaN   
1        18            9  Senior Executive (C-Suite, VP, etc.)   
2        27           23                   Developer, back-end   
3        12            7                  Developer, front-end   
4         6            4                 Developer, full-stack   

                    OrgSize                 PurchaseInfluence      TechList  \
0                       NaN                               NaN           NaN   
1          2 to 9 employees  I have a great deal of influence   Investigate   
2  5,000 to 9,999 employees             I have some influence  Given a list   
3      100 to 499 employees             I have some influence   Investigate   
4        20 to 99 employees             I have some influence   Investigate   

                                          BuyNewTool  \
0                                                NaN   
1  Start a free trial;Ask developers I know/work ...   
2  Start a free trial;Ask developers I know/work ...   
3  Start a free trial;Ask developers I know/work ...   
4  Start a free trial;Ask developers I know/work ...   

                    Country                   Currency  CompTotal  \
0                       NaN                        NaN        NaN   
1  United States of America  USD\tUnited States dollar   285000.0   
2  United States of America  USD\tUnited States dollar   250000.0   
3  United States of A

In [84]:
# This Data Set doesn't seem useful to me. We will disreguard it for now. 
df2.head()

qid     qname                                           question  \
0   QID16        S0  <div><span style="font-size:19px;"><strong>Hel...   
1   QID12  MetaInfo                                  Browser Meta Info   
2  QID310      Q310  <div><span style="font-size:19px;"><strong>You...   
3  QID312      Q120                                                      
4    QID1        S1  <span style="font-size:22px; font-family: aria...   

  force_resp  type selector  
0      False    DB       TB  
1      False  Meta  Browser  
2      False    DB       TB  
3       True    MC     SAVR  
4      False    DB       TB

In [85]:
df.shape

(89184, 84)

In [86]:
df.columns

Index(['ResponseId', 'Q120', 'MainBranch', 'Age', 'Employment', 'RemoteWork',
       'CodingActivities', 'EdLevel', 'LearnCode', 'LearnCodeOnline',
       'LearnCodeCoursesCert', 'YearsCode', 'YearsCodePro', 'DevType',
       'OrgSize', 'PurchaseInfluence', 'TechList', 'BuyNewTool', 'Country',
       'Currency', 'CompTotal', 'LanguageHaveWorkedWith',
       'LanguageWantToWorkWith', 'DatabaseHaveWorkedWith',
       'DatabaseWantToWorkWith', 'PlatformHaveWorkedWith',
       'PlatformWantToWorkWith', 'WebframeHaveWorkedWith',
       'WebframeWantToWorkWith', 'MiscTechHaveWorkedWith',
       'MiscTechWantToWorkWith', 'ToolsTechHaveWorkedWith',
       'ToolsTechWantToWorkWith', 'NEWCollabToolsHaveWorkedWith',
       'NEWCollabToolsWantToWorkWith', 'OpSysPersonal use',
       'OpSysProfessional use', 'OfficeStackAsyncHaveWorkedWith',
       'OfficeStackAsyncWantToWorkWith', 'OfficeStackSyncHaveWorkedWith',
       'OfficeStackSyncWantToWorkWith', 'AISearchHaveWorkedWith',
       'AISearchWan

In [87]:
df.dtypes

ResponseId                               int64
Q120                                    object
MainBranch                              object
Age                                     object
Employment                              object
RemoteWork                              object
CodingActivities                        object
EdLevel                                 object
LearnCode                               object
LearnCodeOnline                         object
LearnCodeCoursesCert                    object
YearsCode                               object
YearsCodePro                            object
DevType                                 object
OrgSize                                 object
PurchaseInfluence                       object
TechList                                object
BuyNewTool                              object
Country                                 object
Currency                                object
CompTotal                              float64
LanguageHaveW

In [88]:
# Check for Null Values. 
pd.set_option('display.max_rows',84) #So we can see all the columns
df.isna().sum()

ResponseId                                 0
Q120                                       0
MainBranch                                 0
Age                                        0
Employment                              1286
RemoteWork                             15374
CodingActivities                       15420
EdLevel                                 1211
LearnCode                               1521
LearnCodeOnline                        19100
LearnCodeCoursesCert                   52108
YearsCode                               1749
YearsCodePro                           23048
DevType                                12312
OrgSize                                24141
PurchaseInfluence                      24220
TechList                               28333
BuyNewTool                              6175
Country                                 1211
Currency                               23850
CompTotal                              40959
LanguageHaveWorkedWith                  2044
LanguageWa

## Data cleaning 🧴 🧼

In [89]:
# Drop duplicate rows (if there are any)
df.drop_duplicates()

ResponseId     Q120                      MainBranch  \
0               1  I agree                   None of these   
1               2  I agree  I am a developer by profession   
2               3  I agree  I am a developer by profession   
3               4  I agree  I am a developer by profession   
4               5  I agree  I am a developer by profession   
...           ...      ...                             ...   
89179       89180  I agree  I am a developer by profession   
89180       89181  I agree  I am a developer by profession   
89181       89182  I agree     I code primarily as a hobby   
89182       89183  I agree  I am a developer by profession   
89183       89184  I agree  I am a developer by profession   

                      Age                                         Employment  \
0         18-24 years old                                                NaN   
1         25-34 years old                                Employed, full-time   
2         45-54 years old                                Employed, full-time   
3         25-34 years old                                Employed, full-time   
4         25-34 years old  Employed, full-time;Independent contractor, fr...   
...                   ...                                                ...   
89179     25-34 years old  Employed, full-time;Independent contractor, fr...   
89180     18-24 years old             Student, full-time;Employed, part-time   
89181   Prefer not to say                                I prefer not to say   
89182  Under 18 years old             Employed, part-time;Student, part-time   
89183     35-44 years old                                Employed, full-time   

                                 RemoteWork  \
0                                       NaN   
1                                    Remote   
2      Hybrid (some remote, some in-person)   
3      Hybrid (some remote, some in-person)   
4                                    Remote   
...                                     ...   
89179                                Remote   
89180  Hybrid (some remote, some in-person)   
89181                                   NaN   
89182  Hybrid (some remote, some in-person)   
89183  Hybrid (some remote, some in-person)   

                                        CodingActivities  \
0                                                    NaN   
1      Hobby;Contribute to open-source projects;Boots...   
2      Hobby;Professional development or self-paced l...   
3                                                  Hobby   
4      Hobby;Contribute to open-source projects;Profe...   
...                                                  ...   
89179  Hobby;Bootstrapping a business;Freelance/contr...   
89180                            School or academic work   
89181                                                NaN   
89182                      Hobby;School or academic work   
89183  Hobby;Professional development or self-paced l...   

                                                 EdLevel  \
0                                                    NaN   
1           Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
2           Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
3           Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
4           Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
...                                                  ...   
89179                Associate degree (A.A., A.S., etc.)   
89180       Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
89181                                     Something else   
89182  Secondary school (e.g. American high school, G...   
89183       Bachelor’s degree (B.A., B.S., B.Eng., etc.)   

                                               LearnCode  \
0                                                    NaN   
1      Books / Physical media;Colleague;Friend or fam...   
2      Books / Physical media;Colleague;On the job tr...   
3      Colleague;Friend or family member;Other online.

### Replacing Null Values

Since a we aren't doing a whole lot of intence data cleaning, we are going to replace the null values with 'Unknown'.

In [90]:
df.fillna("Unknown", inplace = True) # replaceing all null values

In [91]:
df.head(20) # Checking to see if the data set worked. 

ResponseId     Q120                                         MainBranch  \
0            1  I agree                                      None of these   
1            2  I agree                     I am a developer by profession   
2            3  I agree                     I am a developer by profession   
3            4  I agree                     I am a developer by profession   
4            5  I agree                     I am a developer by profession   
5            6  I agree                     I am a developer by profession   
6            7  I agree                     I am a developer by profession   
7            8  I agree                     I am a developer by profession   
8            9  I agree  I am not primarily a developer, but I write co...   
9           10  I agree                     I am a developer by profession   
10          11  I agree                     I am a developer by profession   
11          12  I agree                     I am a developer by profession   
12          13  I agree                     I am a developer by profession   
13          14  I agree  I am not primarily a developer, but I write co...   
14          15  I agree                     I am a developer by profession   
15          16  I agree                     I am a developer by profession   
16          17  I agree                     I am a developer by profession   
17          18  I agree                     I am a developer by profession   
18          19  I agree                     I am a developer by profession   
19          20  I agree                     I am a developer by profession   

                   Age                                         Employment  \
0      18-24 years old                                            Unknown   
1      25-34 years old                                Employed, full-time   
2      45-54 years old                                Employed, full-time   
3      25-34 years old                                Employed, full-time   
4      25-34 years old  Employed, full-time;Independent contractor, fr...   
5      35-44 years old                                Employed, full-time   
6      35-44 years old                                Employed, full-time   
7      25-34 years old                                Employed, full-time   
8      45-54 years old                                Employed, full-time   
9      25-34 years old                 Not employed, but looking for work   
10     25-34 years old                                Employed, full-time   
11     25-34 years old                                Employed, full-time   
12     25-34 years old                                Employed, full-time   
13     35-44 years old                                Employed, full-time   
14     25-34 years old                                Employed, full-time   
15     35-44 years old                                Employed, full-time   
16     35-44 years old                                Employed, full-time   
17     25-34 years old                                Employed, full-time   
18  Under 18 years old                                Employed, full-time   
19     35-44 years old                                Employed, full-time   

                              RemoteWork  \
0                                Unknown   
1                                 Remote   
2   Hybrid (some remote, some in-person)   
3   Hybrid (some remote, some in-person)   
4                                 Remote   
5                                 Remote   
6                                 Remote   
7                                 Remote   
8   Hybrid (some remote, some in-person)   
9                                Unknown   
10                                Remote   
11                                Remote   
12                                Remote   
13  Hybrid (some remote, some in-person)   
14  Hybrid (some remote, some in-person)   
15  Hybrid (some remote, some in-person)   

In [92]:
# Since we don't know the value we will replace it with 1. 
df['ConvertedCompYearly'] = df['ConvertedCompYearly'].replace("Unknown", 1)

In [93]:
df['ConvertedCompYearly'].astype(int)

0             1
1        285000
2        250000
3        156000
4         23456
          ...  
89179         1
89180         1
89181         1
89182         1
89183         1
Name: ConvertedCompYearly, Length: 89184, dtype: int64

In [94]:
df.isna().sum() # There are no longer null values. 

ResponseId                             0
Q120                                   0
MainBranch                             0
Age                                    0
Employment                             0
RemoteWork                             0
CodingActivities                       0
EdLevel                                0
LearnCode                              0
LearnCodeOnline                        0
LearnCodeCoursesCert                   0
YearsCode                              0
YearsCodePro                           0
DevType                                0
OrgSize                                0
PurchaseInfluence                      0
TechList                               0
BuyNewTool                             0
Country                                0
Currency                               0
CompTotal                              0
LanguageHaveWorkedWith                 0
LanguageWantToWorkWith                 0
DatabaseHaveWorkedWith                 0
DatabaseWantToWo

## Analysis of Data


### 1. What jobs paid the most at?

In order to do this we are going to find the Average amount on money people get paid, grouped by the "DevType". 


In [111]:
# We must use a filtered df to not skew the numbers.
filt_df = df[df['DevType'] != "Unknown"] # filtering out unknown jobs
filtered_df = filt_df[filt_df['ConvertedCompYearly'] != 1] # filtering out 1s

In [112]:
# Group by 'DevType' and calculate the average salary for each group
average_salary_by_devtype = filtered_df.groupby('DevType')['ConvertedCompYearly'].mean().sort_values(ascending = False).reset_index()

In [118]:
# Loop through the groups and print the average salary
counter = 0
print('The top 5 most high-playing by dev type:')
for index, row in average_salary_by_devtype.iterrows():
    devtype = row['DevType']
    avg_salary = row['ConvertedCompYearly']
    if counter >= 5: # we only want the top 5 jobs
        break
    print(f"Job Title: {devtype}, Average Yearly Salary: {avg_salary:.2f}")
    counter += 1


The top 5 most high-playing by dev type:
Job Title: Developer Experience, Average Yearly Salary: 301328.45
Job Title: Blockchain, Average Yearly Salary: 170566.61
Job Title: Senior Executive (C-Suite, VP, etc.), Average Yearly Salary: 169683.88
Job Title: Engineering manager, Average Yearly Salary: 153061.82
Job Title: Engineer, site reliability, Average Yearly Salary: 146705.48


#### Conclution: 
Looks like an **Experienced Developer** makes the most money, with an average  yearly salary of: **$301,328.45** 


### 2. How much does remote working matter to employees?

### 3. How does coding experience affect the level of pay?

### 4. What's the most popular method of learning to code?

### 5. Are the chances of getting a job higher as a developer if you have a master's Degree?